### DISCUSSION TITLE
Easyeffects 6.2.8 crackling on headphones when using noise reduction

### DISCUSSION THREAD

**Discussion by yyportne** [wwmm/easyeffects#1734] [General]
Hi, I have started using easyeffects yesterday. I have since noticed that enabling the noise reduction on my microphone using easyeffects causes all the audio output on my phones to crackle. Disabling the noise reduction causes the crackle sounds to disappear. 
Can anyone give me a pointer what I have to tweak to fix these issues?

**Comment by wwmm** [comment]
Does only the noise reduction plugin cause it? Depending on the answer the source of the problem will be different. Assuming it also happens to others the problem may be that the current latency has become too low for you system to handle. In many situations PipeWire may switch to a `too low` latency once you start to use your microphone while applying effects at the same time. I have seen this in my computers a few times. In my desktop nothing unusual happens. Maybe because my `Ryzen 3700x` has plenty of processing power or because my desktop's combination of CPU and soundcard can handle low latency. But in my laptop it is a "crackling festival" when PipeWire does this. The only workaround I have found so far is temporarily force PipeWire to a higher latency when I am using my laptop microphone.

In PipeWire's wiki there is a `quantum-ranges` section hat explains how to temporarily set a custom `quantum size (latency)` https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Config-PipeWire#quantum-ranges. You can see the current quantum used by your microphone and soundcard running `pw-top` while an audio application is recording/playing. If you see something like `256` or lower try to force `1024` or `2048`. The latency will increase. But usually this fixes the crackling if too low latency is what causing it.

**Reply by yyportne** [reply]
Hi thanks for the elaborate answer. For now I can say that the problem is caused only by the noise removal effect. After disabling the effect the crackling goes away, enabling the effect causes crackling again. Other effects do not have this issue. I will look into the quantum ranges configuration and give you feedback later if it worked. Thank you!

**Reply by wwmm** [reply]
Ok. The noise reduction plugin is one of the few where resampling is necessary because the neural network in the `rnnoise` library was trained at `48 kHz`. So maybe the crackling is somehow related to the resampling and not the latency.

**Reply by yyportne** [reply]
I changed the pipewire.conf values to
default.clock.quantum = 1024 and default.clock.min-quantum = 1024
That actually fixed the crackling problem completely. However my audio/video player does not like these values and the audio is super quiet and partially missing. I guess I need to set the quantum value for the noise reduction effect only. Not sure how to do this though at the moment. I need to read the docs you provided me with.

**Reply by wwmm** [reply]
>However my audio/video player does not like these values and the audio is super quiet and partially missing.

That is strange. Increasing the latency shouldn't make the audio quieter. And a quantum size of `1024` is very common and it is not so high. You can calculate the server latency dividing the quantum size by the sampling rate. If the server was at `48 kHz` a quantum size of `1024` will produce a latency of `1024 / 48000 = 0.021 seconds = 21 milliseconds`. This is a typical everyday latency value and should not have caused this kind of side effects.

**Reply by yyportne** [reply]
Alright I checked the pipewire connections and found out that I somehow routed my audio player outputs through the noise reduction. That's why the music loudness was reduced partially. It all works now with correct routing and the 1024 quantum size. Thanks a lot!
